OpenRA.Mods.Common/EditorBrushes/EditorTileBrush.cs

```diff
@@ -106,8 +124,80 @@ public bool HandleMouseInput(MouseInput mi)
 					}
 				}
 			}
+		}
 
-			return true;
+		void FloodFillWithBrush(CPos cell, bool isMoving)
+		{
+			var map = world.Map;
+			var mapTiles = map.MapTiles.Value;
+			var replace = mapTiles[cell];
+
+			if (replace.Type == Template)
+				return;
+
+			var queue = new Queue<CPos>();
+			var touched = new bool[map.MapSize.X, map.MapSize.Y];
+
+			var rules = map.Rules;
+			var tileset = rules.TileSets[map.Tileset];
+			var template = tileset.Templates[Template];
+
+			Action<CPos> maybeEnqueue = (newCell) =>
+			{
+				if (map.Contains(cell) && !touched[newCell.X, newCell.Y])
+				{
+					queue.Enqueue(newCell);
+					touched[newCell.X, newCell.Y] = true;
+				}
+			};
+			Func<CPos, bool> shouldPaint = (cellToCheck) =>
+			{
+				for (var y = 0; y < template.Size.Y; y++)
+				{
+					for (var x = 0; x < template.Size.X; x++)
+					{
+						var c = cellToCheck + new CVec(x, y);
+						if (!map.Contains(c) || mapTiles[c].Type != replace.Type)
+							return false;
+					}
+				}
+
+				return true;
+			};
+
+			Func<CPos, CVec, CPos> findEdge = (refCell, direction) =>
+			{
+				for (;;)
+				{
+					var newCell = refCell + direction;
+					if (!shouldPaint(newCell))
+						return refCell;
+					refCell = newCell;
+				}
+			};
+
+			queue.Enqueue(cell);
+			while (queue.Count > 0)
+			{
+				var queuedCell = queue.Dequeue();
+				if (!shouldPaint(queuedCell))
+					continue;
+
+				var previousCell = findEdge(queuedCell, new CVec(-1 * template.Size.X, 0));
+				var nextCell = findEdge(queuedCell, new CVec(1 * template.Size.X, 0));
+
+				for (var x = previousCell.X; x <= nextCell.X; x += template.Size.X)
+				{
+					PaintCell(new CPos(x, queuedCell.Y), isMoving);
+					var upperCell = new CPos(x, queuedCell.Y - (1 * template.Size.Y));
```

Comment:
I've tested the map editor with the TS mod and it seems to work now, should I change them nevertheless?